src/chainrules.jl

```diff
@@ -11,75 +11,131 @@
     (c, -c, z),
 )
 
+# StatsFuns: https://github.com/JuliaStats/StatsFuns.jl/pull/106
+
 ## Beta ##
 
 @scalar_rule(
     betalogpdf(α::Real, β::Real, x::Number),
-    @setup(di = digamma(α + β)),
+    @setup(z = digamma(α + β)),
     (
-        @thunk(log(x) - digamma(α) + di),
-        @thunk(log(1 - x) - digamma(β) + di),
-        @thunk((α - 1)/x + (1 - β)/(1 - x)),
+        log(x) + z - digamma(α),
+        log1p(-x) + z - digamma(β),
+        (α - 1) / x + (1 - β) / (1 - x),
     ),
 )
 
 ## Gamma ##
 
 @scalar_rule(
     gammalogpdf(k::Real, θ::Real, x::Number),
+    @setup(
+        invθ = inv(θ),
+        xoθ = invθ * x,
+        z = xoθ - k,
+    ),
     (
-        @thunk(-digamma(k) - log(θ) + log(x)),
-        @thunk(-k/θ + x/θ^2),
-        @thunk((k - 1)/x - 1/θ),
+        log(xoθ) - digamma(k),
+        invθ * z,
+        - (1 + z) / x,
     ),
 )
 
 ## Chisq ##
 
 @scalar_rule(
     chisqlogpdf(k::Real, x::Number),
-    @setup(ko2 = k / 2),
-    (@thunk((-logtwo - digamma(ko2) + log(x)) / 2), @thunk((ko2 - 1)/x - one(ko2) / 2)),
+    @setup(hk = k / 2),
+    (
+        (log(x) - logtwo - digamma(hk)) / 2,
```

Comment:
Where is `logtwo` from? I see it's from StatsFuns, but I can't find the definition in StatsFuns :confused: 

I ask because I worry it might lead to undeseriable type-promotion, e.g. `Float32` to `Float64`.

Reply:
It's defined in LogExpFunctions and just reexported from StatsFuns (the PR was finally merged some days ago :tada:). It is defined as an `Irrational` and hence should avoid type promotions if possible, e.g:
```julia
julia> logtwo - 3.4
-2.7068528194400545

julia> logtwo - 3.4f0
-2.706853f0
```

Reply:
Good, good :+1: Yeah in this case it should be fine, but there are several cases in DiffRules.jl (and I'm pretty certain the same ones exist in ChainRules.jl?) where the operations are written in a way such that `Irrational` is first converted into `Float64`, and thus the dual is promoted to `Float64`, despite the primal being `Float32` (https://github.com/JuliaDiff/DiffRules.jl/pull/55).

And the annoying bit is that some of these constants are defined in StatsFuns.jl, some in LogExpFunctions.jl, etc., so it's difficult to re-use them in DiffRules.jl :confused:
Though maybe I should actually just add all those constants to that PR instead of doing `oftype` everywhere, hmm.

Reply:
It's only a problem if you perform operations with the constants (such as `sqrt`) before they are promoted to the correct type, which is not the case here.